### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by TasbeehTakrori**

### CODE REVIEW

File: WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+using WeatherMonitoringAndReportingService.BotFactory;
+using WeatherMonitoringAndReportingService.Models;
+
+namespace WeatherService.Test.BotFactory
+{
+    public class WeatherBotFactoryTests
+    {
+        [Theory]
+        [AutoMoqData]
+        public void GetEnabledWeatherObservers_InvalidBotSettings_ThrowException(
```
Comment (kshawish) [inline WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs:14]: Note for myself to discuss what are we testing here.

File: WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    Enabled = false,
+                    TemperatureThreshold = 0,
+                    Message = "Brrr, it's getting chilly!"
+                });
+
+            // Act
+            var observers = sut.GetEnabledWeatherObservers();
+
+            // Assert
+            observers.Should().HaveCount(2);
```
Comment (kshawish) [inline WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs:59]: is it enough to check only the count?

File: WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+            //Act & Assert
+            Action Act = () => sut.GetEnabledWeatherObservers();
+            Act.Should().Throw<Exception>().WithMessage(
+                "Failed to create bots. An error occurred while initializing bot instances.");
+        }
+
+        [Theory]
+        [AutoMoqData]
+        public void GetEnabledWeatherObservers_Should_ReturnListOfIWeatherObserver(
```
Comment (kshawish) [inline WeatherService.Test/BotFactory/WeatherBotFactoryTests.cs:28]: there's a problem with this test, that if you added another bot without changing the behavior of GetEnabledWeatherObservers the test would fail.

File: WeatherService.Test/BotSettingsReaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+using WeatherMonitoringAndReportingService;
+
+namespace WeatherService.Test
+{
+    public class BotSettingsReaderTests
+    {
+        [Theory]
+        [InlineData("RainBot")]
+        [InlineData("SunBot")]
+        public void GetBotSettings_Should_ReturnBotSettings(string key)
```
Comment (kshawish) [inline WeatherService.Test/BotSettingsReaderTests.cs:12]: note for myself to discuss what are we testing here.

File: WeatherService.Test/BotSettingsReaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                .SetBasePath(AppDomain.CurrentDomain.BaseDirectory)
+                .AddJsonFile("appsettings.json")
+                .Build();
+            var botSettingsReader = new BotSettingsReader(configuration);
+
+            //Act
+            var botSettings = botSettingsReader.GetBotSettings(key);
+
+            //Assert
+            botSettings.Should().NotBeNull();
```
Comment (kshawish) [inline WeatherService.Test/BotSettingsReaderTests.cs:25]: is this enough to only assert the the result is not null ?

File: WeatherService.Test/ConsoleOutputCapture.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿public class ConsoleOutputCapture : IDisposable
+{
+    private readonly StringWriter consoleOutput;
+
+    public ConsoleOutputCapture()
```
Comment (kshawish) [inline WeatherService.Test/ConsoleOutputCapture.cs:5]: ok that's new to me, sound good. but If we care about testing a service that uses console, to have it depend on an abstraction that use the console rather than using it directly.

File: WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace WeatherService.Test.DataHandlingTests.DataParsingTests
+{
+    public class DataParserTests
+    {
+        [Theory]
+        [AutoMoqData]
+        public void ParseData_NoSuitableStrategy_ThrowsException(
+          [Frozen] Mock<IDataParsingStrategy> mockStrategy,
+          string invalidData,
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs:15]: would prefer to have a static invalid data.

File: WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+using WeatherMonitoringAndReportingService.Models;
+
+namespace WeatherService.Test.DataHandlingTests.DataParsingTests
+{
+    public class DataParserTests
+    {
+        [Theory]
+        [AutoMoqData]
+        public void ParseData_NoSuitableStrategy_ThrowsException(
+          [Frozen] Mock<IDataParsingStrategy> mockStrategy,
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs:14]: Let's discuss how this works specially that DataParser depends on `List< IDataParsingStrategy>`no just a `IDataParsingStrategy`

File: WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+            // Act and Assert
+            Action act = () => sut.ParseData(invalidData);
+            act.Should().Throw<Exception>()
+                .WithMessage("Failed to determine a suitable parsing strategy for the provided data.");
+        }
+
+        [Theory]
+        [AutoMoqData]
+        public void ParseData_ValidData_ReturnsParsedData(
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/DataParserTests.cs:29]: nice, but let's discuss it to make sure you totally understand what are you testing here.

File: WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        [InlineAutoData("{\"Location\": \"City Name\", \"Temperature\": 32, \"Humidity\": 40}",
+            "City Name", 32, 40)]
+        public void ParseData_Should_ReturnWeatherData(
+            string data, string expectedLocation, int expectedTemperature, int expectedHumidity, JsonParsingStrategy sut)
+        {
+            // Act
+            WeatherData weatherData = sut.ParseData(data);
+
+            // Assert
+            weatherData.Location.Should().Be(expectedLocation);
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs:21]: check if there's a way to use FluentAssertion to assert on the whole object rather than doing that for each field.

File: WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            weatherData.Temperature.Should().Be(expectedTemperature);
+            weatherData.Humidity.Should().Be(expectedHumidity);
+        }
+
+        [Theory]
+        [AutoData]
+        public void ParseData_Should_ThrowException_WhenParsingFailed(JsonParsingStrategy sut)
+        {
+            // Act and Assert
+            Action act = () => sut.ParseData(It.IsAny<string>());
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs:31]: let's use a literal invalid Json rather than `It.IsAny<string>()`

File: WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            //Assert
+            isDataCompatible.Should().BeTrue();
+        }
+
+        [Theory]
+        [AutoData]
+        public void IsDataCompatible_Should_ReturnFalse(JsonParsingStrategy sut)
+        {
+            //Act
+            var isDataCompatible = sut.IsDataCompatible("");
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/JsonParsingStrategyTests.cs:51]: ok nice, but empty string might act as an edge case that need it's own test case (as null). would explain.

File: WeatherService.Test/DataHandlingTests/DataParsingTests/XmlParsingStrategyTests.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
+﻿using AutoFixture.Xunit2;
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataParsingTests/XmlParsingStrategyTests.cs:1]: same comments as the JsonParser.

File: WeatherService.Test/DataHandlingTests/DataReaderTests.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+            //Arrange
+            using (var fakeInputReader = new StringReader(fakeInput))
+            {
+                Console.SetIn(fakeInputReader);
+
+                //Act
+                var rawData = sut.ReadRawData();
+
+                //Assert
+                rawData.Should().NotBeNullOrWhiteSpace();
```
Comment (kshawish) [inline WeatherService.Test/DataHandlingTests/DataReaderTests.cs:22]: is this assertion enough to make sure that the `ReadRawData` work properly ?

File: WeatherService.Test/StartupTests.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+using WeatherMonitoringAndReportingService.DataHandling.DataParsing;
+using WeatherMonitoringAndReportingService.WeatherReportPublishing;
+
+namespace WeatherService.Test
+{
+    public class StartupTests
+    {
+        [Theory]
+        [AutoMoqData]
+        public void InitializeWeatherMonitoringService_Should_ReturnInitializationObjects(
```
Comment (kshawish) [inline WeatherService.Test/StartupTests.cs:15]: let's discuss this, can't see any value of it.

File: WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+            //Arrange
+            botSettings.HumidityThreshold = 70;
+            weatherData.Humidity = 50;
+
+            //Act
+            sut.Run(weatherData);
+            var captureOutput = consoleOutputCapture.GetCapturedOutput();
+            //Assert
+            captureOutput.Should().BeEmpty();
+            captureOutput.Should().NotContain("RainBot activated!");
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs:43]: no need since you check that it's empty

File: WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            var captureOutput = consoleOutputCapture.GetCapturedOutput();
+            //Assert
+            captureOutput.Should().NotBeNull();
+            captureOutput.Should().Contain("RainBot activated!");
+        }
+
+        [Theory]
+        [AutoData]
+        public void Run_ShouldNotActivateBot_WhenHumidityIsLessThanThreshold(
+            [Frozen] BotSettings botSettings, WeatherData weatherData,
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs:31]: what does [Frozen] do here ?

File: WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+            ConsoleOutputCapture consoleOutputCapture, RainBot sut)
+        {
+            //Arrange
+            botSettings.HumidityThreshold = 70;
+            weatherData.Humidity = 50;
+
+            //Act
+            sut.Run(weatherData);
+            var captureOutput = consoleOutputCapture.GetCapturedOutput();
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/RainBotTests.cs:40]: it's not explicit how this read the console output, let's talk about it

File: WeatherService.Test/WeatherReportPublishingTests/SnowBotTests.cs [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+        public void Run_ShouldActivateBot_WhenTemperatureIsLessThanThreshold(
+            [Frozen] BotSettings botSettings, WeatherData weatherData,
+            SnowBot sut)
+        {
+            //Arrange
+            botSettings.TemperatureThreshold = 0;
+            weatherData.Temperature = -2;
+
+            var writer = new StringWriter();
+            Console.SetOut(writer);
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/SnowBotTests.cs:21]: why you don't use ConsoleOutputCapture as other cases ?

File: WeatherService.Test/WeatherReportPublishingTests/SnowBotTests.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+            //Arrange
+            botSettings.TemperatureThreshold = 0;
+            weatherData.Temperature = 10;
+
+            //Act
+            sut.Run(weatherData);
+            var captureOutput = consoleOutputCapture.GetCapturedOutput();
+            //Assert
+            captureOutput.Should().BeEmpty();
+            captureOutput.Should().NotContain("SnowBot activated!");
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/SnowBotTests.cs:46]: no need

File: WeatherService.Test/WeatherReportPublishingTests/SunBotTests.cs [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+            //Arrange
+            botSettings.TemperatureThreshold = 40;
+            weatherData.Temperature = 30;
+
+            //Act
+            sut.Run(weatherData);
+            var captureOutput = consoleOutputCapture.GetCapturedOutput();
+            //Assert
+            captureOutput.Should().BeEmpty();
+            captureOutput.Should().NotContain("SunBot activated!");
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/SunBotTests.cs:43]: no need.

File: WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+            WeatherReportPublisher sut)
+        {
+            //Arrange
+            sut.Subscribe(mockObserver.Object);
+
+            //Act
+            sut.Unsubscribe(mockObserver.Object);
+
+            //Assert
+            sut.GetSubscribers().Should().NotContain(mockObserver.Object);
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs:52]: let's discuss how we test this in a different way without depending on GetSubscribers

File: WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+            WeatherReportPublisher sut)
+        {
+            //Arrange
+            sut.Subscribe(mockObserver.Object);
+
+            //Act
+            sut.NotifySubscribers(weatherData);
+
+            //Assert
+            mockObserver.Verify(x => x.Run(weatherData), Times.Exactly(sut.GetSubscribers().Count()));
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs:69]: the subscriber should run only once.

File: WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+        public void PublishData_Should_InvokeNotifySubscribers(
+            Mock<IWeatherObserver> mockObserver,
+            WeatherData weatherData,
+            WeatherReportPublisher sut)
+        {
+            // Arrange
+            sut.Subscribe(mockObserver.Object);
+
+            // Act
+            sut.PublishData(weatherData);
```
Comment (kshawish) [inline WeatherService.Test/WeatherReportPublishingTests/WeatherReportPublisherTests.cs:83]: not sure what's the different between publish and Notify subscribers

**Review by kshawish — COMMENTED** [review]
Good Job Tasbeeh
